### PR TITLE
#975 new tool pyRevit Health Check

### PR DESCRIPTION
**Pull request by davidvadkerti**
I've finally found some time to add the Model Checker tool.
Unfortunatelly I am sure few things are not done the best way. I wrote them also in the comments:

1. hardcoded list of colors for charts instead of _randomize_colors()_
2. error when charts render first time
3. treating errors when chart data has accented characters - I just tried to remove them by _accents2ascii_ function. I personally think you shouldn't use accented characters in the name of worksets but unfortunatelly in my experience it happens. 
4. help article for bad revit practices - maybe you know more appropriate article
5. tresholds for parameters are maybe subjective a should be rather configurable?

### CODE REVIEW

**Comment by jmcouffin** [comment]
The fact that the threshold are hardcoded is an issue. They are very subjective.
It needs to be stored / setup per project or per specs: 

- big, 
- medium,
-  small, 
- tower projects, 
- house projects

Maybe in a separate process. One that you need to launch to setup the threshold choice once for the project. and stored in the project file as a parameter like ThresholdSpecId.

**Comment by jmcouffin** [comment]
another thing that needs some fixing in the code: you are using the lookup parameter function twice, therefore some of the model check do not work properly when using Revit in another language.

**Comment by davidvadkerti** [comment]
Ok, I will try to fix these issues. Thanks for the review :)

**Comment by jmcouffin** [comment]
I would have fixed it, but did not have time. Excellent tool Davide.

Le sam. 24 oct. 2020 12:24, davidvadkerti <notifications@github.com> a
écrit :

**Comment by jmcouffin** [comment]
check that in the remarks section.
https://www.revitapidocs.com/2015/4400b9f8-3787-0947-5113-2522ff5e5de2.htm
and this might help as well
https://spiderinnet.typepad.com/blog/2020/05/revit-net-api-get-parameters-from-element.html
and this https://forums.autodesk.com/t5/revit-api-forum/get-parameter-guid-or-lookupparameter-whitch-is-better/td-p/6453524
I tend to use get_Parameter(_BuiltinParameterName_) method. bullet proof, and does not vary from one language to the other.
I have learnt quite a few tricks looking at your code, nice job!

**Comment by jmcouffin** [comment]
line 138:
`     viewName = view.GetParameters('VIEW_NAME")`
line 348:
`     widthFactor = textnote.GetParameters("TEXT_WIDTH_SCALE")`

**Comment by davidvadkerti** [comment]
Thanks for the help, I will try to finally fix it this week.

**Comment by jmcouffin** [comment]
I just edited the line numbers. I messed up ;p
I still have issues with the draw method for charts with my french sample model
`chartCategories.draw()`
and
`chartFCategories.draw()`
It has to do with accents but I don't know precisely where it is not well handled

**Comment by davidvadkerti** [comment]
I have issue drawing the Chart.js chart with any accented values. I think there should be some smarter workaround but I was just happy using my crappy workflow with function removing accents - `accents2ascii(string)` - line 84

Try to uncomment line 423 if it makes any change:
`category = accents2ascii(category)` .
I was using it  just on revit with english GUI so I havent come across accents in `Category.Name` yet.

Another issue is filtering out categories which I find not usefull to show. I should probably rather compare BuiltInCategories instead of Category.Names. - line 414

**Comment by jmcouffin** [comment]
yep the accents issue is huge, the names of parameters also., uncommenting did not work

**Comment by jmcouffin** [comment]
a `text = text.encode(‘utf-8’) `would not do the job? (I didn't try it)

**Comment by jmcouffin** [comment]
`graphCatHeadings = [x.encode('UTF8') for x in graphCatHeadings]`
add this piece to the puzzle after the graphCatHeadings creation

**Comment by jmcouffin** [comment]
and 
```
graphFCatData = [x.encode('utf8') for x in graphFCatData]
graphFCatHeadings = [x.encode('utf8') for x in graphFCatHeadings]
```
after the for loop on # INPLACE CATEGORY GRAPH

File: extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+print("\n\n\n\n")
+
+# INPLACE CATEGORY GRAPH
+fCatSet=[]
+# sorting results in chart legend
+graphFCatHeadings.sort()
+for i in graphFCatHeadings:
+    count=graphFCatData.count(i)        
+    fCatSet.append(count)
+
```
Comment (jmcouffin) [inline extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py:543]:
```
graphFCatData = [x.encode('utf8') for x in graphFCatData]
graphFCatHeadings = [x.encode('utf8') for x in graphFCatHeadings]
```

File: extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+    except:
+        pass
+
+catSet=[]
+# sorting results in chart legend
+graphCatHeadings.sort()
+for i in graphCatHeadings:
+    count=graphCatData.count(i)        
+    catSet.append(count)
+
```
Comment (jmcouffin) [inline extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py:456]: `graphCatHeadings = [x.encode('UTF8') for x in graphCatHeadings]`

File: extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+views_id_collector = FilteredElementCollector(doc).OfCategory(BuiltInCategory.OST_Views) \
+.WhereElementIsNotElementType()
+view_elements = views_id_collector.ToElements()
+viewCount = len(view_elements)
+
+copiedView = 0
+for view in view_elements:
+    viewName = view.LookupParameter('View Name')
+    try:
+        viewNameString = viewName.AsString()
```
Comment (jmcouffin) [inline extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py:169]: `viewNameString = revit.query.get_name(view)`

File: extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+textnoteWFtres = 0
+textnoteCaps = 0
+rampTres = 0
+archTres = 0
+
+# Text notes width factor != 1
+textNoteType_collector = FilteredElementCollector(doc).OfClass(TextNoteType).ToElements()
+textnoteWFcount = 0
+for textnote in textNoteType_collector:
+    widthFactor = textnote.LookupParameter('Width Factor').AsDouble()
```
Comment (jmcouffin) [inline extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py]: `    widthFactor = textnote.get_Parameter(DB.BuiltInParameter.TEXT_WIDTH_SCALE).AsDouble()`

**Review by jmcouffin — CHANGES_REQUESTED** [review]
Added all the stuff I left in the comments. It now works in Revit in French and I am guessing in many other languages

File: extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py [hunk 5]
```diff
@@ -31,19 +31,33 @@
...
-    'Room Tag is outside of its Room',
-    'Multiple Rooms are in the same enclosed region',
-    'Multiple Areas are in the same enclosed region',
-    'One element is completely inside another',
-    'Room is not in a properly enclosed region',
-    'Room separation line is slightly off axis and may cause inaccuracies',
-    'Area is not in a properly enclosed region',
-    "Rectangular opening doesn't cut its host",
-    'Elements have duplicate "Number" values',]
+# citical warnings Guids
```
Comment (davidvadkerti) [inline extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py:34]: I changed Critical warning list to list of GUIDs. Now it is compatible to other laguages than english. I am not sure if there is some better identifier than GUIDs.

File: extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py [hunk 6]
```diff
@@ -411,29 +427,36 @@ def accents2ascii(s):
...
-    'Space Type Settings','Survey Point','Project Base Point','Electrical Demand Factor Definitions','Electrical Load Classifications','Panel Schedule Templates - Branch Panel',
-    'Panel Schedule Templates - Data Panel','Panel Schedule Templates - Switchboard','Electrical Load Classification Parameter Element','Automatic Sketch Dimensions',]
-for i in elements:
+# catBanlist = ['Shared Site','Project Information','Structural Load Cases','Sun Path','Color Fill Schema','HVAC Zones','HVAC Load Schedules','Building Type Settings',
+#     'Space Type Settings','Survey Point','Project Base Point','Electrical Demand Factor Definitions','Electrical Load Classifications','Panel Schedule Templates - Branch Panel',
+#     'Panel Schedule Templates - Data Panel','Panel Schedule Templates - Switchboard','Electrical Load Classification Parameter Element','Automatic Sketch Dimensions',]
+catBanlist = [-2000110,-2003101,-2005210,-2009609,-2000552,-2008107,-2008121,-2008120,-2008119,-2001272,-2001271,-2008142,-2008143,-2008145,-2008147,-2008146,-2008148,-2000261,]
```
Comment (davidvadkerti) [inline extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py:432]: catBanlist - list of categories not show is changed to list of category IDs

File: extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py [hunk 7]
```diff
@@ -475,18 +498,20 @@ def accents2ascii(s):
...
         graphWorksetsData.append(worksetName)
 # print worksetNames
 # sorting results in chart legend
 worksetNames.sort()
+
 worksetsSet=[]
 for i in worksetNames:
     count=graphWorksetsData.count(i)        
     worksetsSet.append(count)
+worksetNames = [x.encode('utf8') for x in worksetNames]
```
Comment (davidvadkerti) [inline extensions/pyRevitTools.extension/pyRevit.tab/Analysis.panel/Tools.stack/Inspect.pulldown/Model Checker.pushbutton/script.py:511]: I added encoding also for worksets, if there was an accent in Workset name it caused an error

**Comment by davidvadkerti** [comment]
> `graphCatHeadings = [x.encode('UTF8') for x in graphCatHeadings]`
> add this piece to the puzzle after the graphCatHeadings creation

I remember I tried x.encode() not in the list comprehension but inside the loop and it didn't work. I must have done something wrong.

**Comment by davidvadkerti** [comment]
> The fact that the threshold are hardcoded is an issue. They are very subjective.
> It needs to be stored / setup per project or per specs:
> 
> * big,
> * medium,
> * small,
> * tower projects,
> * house projects
> 
> Maybe in a separate process. One that you need to launch to setup the threshold choice once for the project. and stored in the project file as a parameter like ThresholdSpecId.

I've tested french, czech, russian and japaneese versions of Revit and it seemed the accents issue should be fixed now. But I can't speak these languages except of czech. I've just looked for glitches in the text.

You mentioned I should store threshold values in the project. I was trying to find out how to create new Shared Parameter using Revit API but haven't succeed yet. I've tried to use rsparam module but I've managed just to create new empty Shared Parameter file. Do you have some reference which could I use?

Thanks for the help and for the patience with me :)

**Comment by jmcouffin** [comment]
That is a great tool that has a great potential and will attract people to pyrevit imho @davidvadkerti 

@eirannejad it should have its own push-button visible in the ui, I think

**Comment by eirannejad** [comment]
Working on merging the PRs today and publishing the v4.8.4

**Comment by eirannejad** [comment]
Merged for review

**Comment by eirannejad** [comment]
@jmcouffin @davidvadkerti Thanks a lot for all the work on this. I merged this for review and ended up combining it with another idea I had that never got around to implement. The **Preflight Checks**

Switch to the `feature/preflight` branch to test this feature.

---

pyRevit extensions, can now define any checks (`*_check.py`) under the `checks/` directory inside the extension. I implemented a new module `pyrevit.preflight` that collects these checks and can orchestrate their execution. This allows the users to create as many checks as they need in their extensions and automatically incorporate that into the default **Preflight Checks** tool. It also allows sharing these checks using pyRevit extensions. The **Preflight Checks** tool will ask the user to run any combination of the available checks.


Run the preflight checks tool from here:

explorer_rL8pOaDNOI

You should see the selection window popup with your test listed in there:


I moved the model checker to `extensions/pyRevitTools.extension/checks/modelcheker_check.py` for testing. This script defines a check case (derived from `pyrevit.preflight.PreflightTestCase`) and implements the necessary interface:

```python
class ModelChecker(PreflightTestCase):
    """Revit model quality check"""

    name = "Model Checker"
    author = "David Vadkerti"

    def setUp(self, doc, output):
        pass

    def startTest(self, doc, output):
        timer = coreutils.Timer()
        checkModel(doc, output)
        endtime = timer.get_time()
        endtime_hms = str(datetime.timedelta(seconds=endtime))
        endtime_hms_claim = "Transaction took " + endtime_hms
        print(endtime_hms_claim)

    def tearDown(self, doc, output):
        pass

    def doCleanups(self, doc, output):
        pass
```
 
I moved all the test logic tentatively to `checkModel()` that is being called inside the `startTest()`

Let's do a good cleanup on this script (split the code into logical functions, comment the code, etc.) and prepare for the feature release 👍

**Comment by eirannejad** [comment]
Let's cleanup the script and I'll improve the Preflight user interface meanwhile. Please checkout this branch, make your changes and send a PR to this branch as well.

**Comment by Robbo1234** [comment]
Would it be possible to add the following Preflightchecks?

**Location:**
Angle to True North
Project Base Point
Survey Point

**Files:**
Keynote
Shared Parameters

**Duplication Marks:**
1. Duplicated Rooms Names
2. Duplicated Window Names
3. Duplicated Door Names

**Areas:**
1. Not Enclosed Areas
2. Not Enclosed Rooms
3. Not Place Areas
4. Not Placed Rooms
5. Placed Areas
6. Placed Room

**Links:**
1. Duplication Cads
2. Imported Cads
3. Linked DGN's
4. Linked DWG's
5. Linked Images
6. PDF Instance
7. PDF's
8. Revit

**Misc**
1. Overridden Dimensions
2. Size of the Revit Model
3. Owners of the Revit Elements in the Model
4. Constraint Elements Count

**Comment by jmcouffin** [comment]
The list in itself makes sense, but it feels like a grocery list anyway.
They are all pretty much possible to implement. Would you help @Robbo1234 ?

**Pull request merged**